crates/cext/src/transpiler/target.rs

```diff
@@ -0,0 +1,1418 @@
+/// are no parameters to be added.
+/// @param property_map The mapping of qargs and InstructionProperties to
+/// be associated with this instruction.
+///
+/// # Example
+///     
+///     QkTarget *target = qk_target_new("New Target", 5);
+///     QkPropsMap *props_map = qk_property_map_new();
+///     uint32_t qargs[2] = {0, 1};
+///     double params[1] = {3.1415};
+///     qk_property_map_add(props_map, qargs, 2, qk_instruction_properties_new(0.0, 0.1));
+///     qk_target_add_instruction(target, QkGate_CRX, params, props_map);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+///
+/// The ``params`` type is expected to be a pointer to an array of ``double`` where the length
+/// matches the the expectations of the standard gate. If the array is insufficently long the
+/// behavior of this function is undefined as this will read outside the bounds of the array.
+/// It can be a null pointer if there are no params for a given gate. You can check
+/// `qk_gate_num_params` to determine how many qubits are required for a given gate.
+///
+/// Behavior is undefined if ``property_map`` is not a valid, non-null pointer to a ``QkPropsMap``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_add_instruction(
+    target: *mut Target,
+    operation: StandardGate,
+    params: *mut f64,
+    property_map: *const PropertyMap,
+) -> ExitCode {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let target = unsafe { mut_ptr_as_ref(target) };
+
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let property_map = unsafe { const_ptr_as_ref(property_map) };
+    // SAFETY: Per the documentation the params pointers are arrays of num_params() elements.
+    let parsed_params: &[Param] = unsafe {
+        match operation.num_params() {
+            0 => &[],
+            1 => &[(*params.wrapping_add(0)).into()],
+            2 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+            ],
+            3 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+                (*params.wrapping_add(2)).into(),
+            ],
+            4 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+                (*params.wrapping_add(2)).into(),
+                (*params.wrapping_add(3)).into(),
+            ],
+            // There are no standard gates that take > 4 params
+            _ => unreachable!(),
+        }
+    };
+    let props_map = if property_map.0.is_empty() {
+        None
+    } else {
+        Some(property_map.0.clone())
+    };
+
+    if target
+        .add_instruction(operation.into(), parsed_params, None, props_map)
+        .is_ok()
+    {
+        ExitCode::Success
+    } else {
+        ExitCode::CInputError
```

Comment:
We can also add a custom error code, `CInputError` is very generically saying that the input from C was faulty. Maybe we can have a transpiler-specific set of error codes with target errors being part of them?

Reply:
Fixed in 46cc693